In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Data loading

In [3]:
file_path = "../datasets/phpB0xrNj.csv"

data = pd.read_csv(file_path)
data.shape

(7797, 618)

In [4]:
data.head()

f1      f2      f3      f4      f5      f6      f7      f8      f9  \
0 -0.4394 -0.0930  0.1718  0.4620  0.6226  0.4704  0.3578  0.0478 -0.1184   
1 -0.4348 -0.1198  0.2474  0.4036  0.5026  0.6328  0.4948  0.0338 -0.0520   
2 -0.2330  0.2124  0.5014  0.5222 -0.3422 -0.5840 -0.7168 -0.6342 -0.8614   
3 -0.3808 -0.0096  0.2602  0.2554 -0.4290 -0.6746 -0.6868 -0.6650 -0.8410   
4 -0.3412  0.0946  0.6082  0.6216 -0.1622 -0.3784 -0.4324 -0.4358 -0.4966   

      f10  ...    f609    f610    f611    f612    f613    f614    f615  \
0 -0.2310  ...  0.4102  0.2052  0.3846  0.3590  0.5898  0.3334  0.6410   
1 -0.1302  ...  0.0000  0.2954  0.2046  0.4772  0.0454  0.2046  0.4318   
2 -0.8318  ... -0.1112 -0.0476 -0.1746  0.0318 -0.0476  0.1112  0.2540   
3 -0.9614  ... -0.0504 -0.0360 -0.1224  0.1366  0.2950  0.0792 -0.0072   
4 -0.5406  ...  0.1562  0.3124  0.2500 -0.0938  0.1562  0.3124  0.3124   

     f616    f617  class  
0  0.5898 -0.4872    '1'  
1  0.4546 -0.0910    '1'  
2  0.1588 -0.4762    '2'  
3  0.0936 -0.1510    '2'  
4  0.2188 -0.2500    '3'  

[5 rows x 618 columns]

In [5]:
data.columns

Index(['f1', 'f2', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'f10',
       ...
       'f609', 'f610', 'f611', 'f612', 'f613', 'f614', 'f615', 'f616', 'f617',
       'class'],
      dtype='object', length=618)

In [6]:
data['class'].unique()

array(["'1'", "'2'", "'3'", "'4'", "'5'", "'6'", "'7'", "'8'", "'9'",
       "'10'", "'11'", "'12'", "'13'", "'14'", "'15'", "'16'", "'17'",
       "'18'", "'19'", "'20'", "'21'", "'22'", "'23'", "'24'", "'25'",
       "'26'"], dtype=object)

## Analysis:
### Total features = 617
### Total unique class = 26
### Total Row = 7797

In [7]:
data.isnull().sum()

f1       0
f2       0
f3       0
f4       0
f5       0
        ..
f614     0
f615     0
f616     0
f617     0
class    0
Length: 618, dtype: int64

In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7797 entries, 0 to 7796
Columns: 618 entries, f1 to class
dtypes: float64(613), int64(4), object(1)
memory usage: 36.8+ MB


In [12]:
data.describe()

f1           f2           f3           f4           f5  \
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000   
mean     -0.385396     0.141103     0.347384     0.450659     0.324283   
std       0.236721     0.320090     0.326658     0.308301     0.461491   
min      -1.000000    -0.892600    -0.975200    -0.968000    -0.996600   
25%      -0.546800    -0.099400     0.105400     0.247800    -0.024800   
50%      -0.414000     0.106400     0.333400     0.453200     0.415800   
75%      -0.241000     0.354600     0.583400     0.668600     0.692400   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                f6           f7           f8           f9          f10  ...  \
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000  ...   
mean      0.163589     0.023852    -0.081577    -0.149191    -0.191763  ...   
std       0.569970     0.559704     0.535142     0.528714     0.515343  ...   
min      -1.000000    -1.000000    -1.000000    -1.000000    -1.000000  ...   
25%      -0.359600    -0.461600    -0.526800    -0.577400    -0.601200  ...   
50%       0.204200    -0.010800    -0.154400    -0.233200    -0.261000  ...   
75%       0.686200     0.492400     0.304400     0.190200     0.120000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              f608         f609         f610         f611         f612  \
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000   
mean      0.370768     0.341730     0.311678     0.318904     0.334451   
std       0.335582     0.330835     0.328813     0.333747     0.334954   
min      -0.609400    -0.803400    -0.775200    -0.830200    -0.711800   
25%       0.130400     0.098000     0.071400     0.073600     0.088600   
50%       0.366200     0.337400     0.294200     0.295000     0.315800   
75%       0.622600     0.588800     0.548800     0.565800     0.581600   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              f613         f614         f615         f616         f617  
count  7797.000000  7797.000000  7797.000000  7797.000000  7797.000000  
mean      0.299238     0.224122     0.186668     0.081370    -0.281416  
std       0.334282     0.336135     0.337780     0.332859     0.354418  
min      -0.770800    -0.810600    -0.780200    -0.802800    -0.962600  
25%       0.049200    -0.017800    -0.057600    -0.154600    -0.546600  
50%       0.280000     0.192400     0.157800     0.045200    -0.329000  
75%       0.541200     0.461600     0.410800     0.289800    -0.065600  
max       1.000000     1.000000     1.000000     1.000000     0.889800  

[8 rows x 617 columns]

In [23]:
targets = data.pop('class')

In [24]:
data.head()

f1      f2      f3      f4      f5      f6      f7      f8      f9  \
0 -0.4394 -0.0930  0.1718  0.4620  0.6226  0.4704  0.3578  0.0478 -0.1184   
1 -0.4348 -0.1198  0.2474  0.4036  0.5026  0.6328  0.4948  0.0338 -0.0520   
2 -0.2330  0.2124  0.5014  0.5222 -0.3422 -0.5840 -0.7168 -0.6342 -0.8614   
3 -0.3808 -0.0096  0.2602  0.2554 -0.4290 -0.6746 -0.6868 -0.6650 -0.8410   
4 -0.3412  0.0946  0.6082  0.6216 -0.1622 -0.3784 -0.4324 -0.4358 -0.4966   

      f10  ...    f608    f609    f610    f611    f612    f613    f614  \
0 -0.2310  ...  0.3334  0.4102  0.2052  0.3846  0.3590  0.5898  0.3334   
1 -0.1302  ...  0.2272  0.0000  0.2954  0.2046  0.4772  0.0454  0.2046   
2 -0.8318  ...  0.0952 -0.1112 -0.0476 -0.1746  0.0318 -0.0476  0.1112   
3 -0.9614  ...  0.0648 -0.0504 -0.0360 -0.1224  0.1366  0.2950  0.0792   
4 -0.5406  ...  0.2812  0.1562  0.3124  0.2500 -0.0938  0.1562  0.3124   

     f615    f616    f617  
0  0.6410  0.5898 -0.4872  
1  0.4318  0.4546 -0.0910  
2  0.2540  0.1588 -0.4762  
3 -0.0072  0.0936 -0.1510  
4  0.3124  0.2188 -0.2500  

[5 rows x 617 columns]

In [25]:
targets.head()

0    '1'
1    '1'
2    '2'
3    '2'
4    '3'
Name: class, dtype: object

In [26]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [28]:
random_seed = 1124

X_train, X_test, y_train, y_test = train_test_split(
    data,
    targets, 
    test_size=0.2,
    random_state=random_seed
)

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(6237, 617)
(1560, 617)
(6237,)
(1560,)


In [29]:
rf_model1 = RandomForestClassifier()

In [30]:
# fit the model with the train data
rf_model1.fit(X_train, y_train)

# prediction from training data
preds_train = rf_model1.predict(X_train)

# accuracy score from the traing data
acc_train = accuracy_score(y_train, preds_train)

# prediction from test data
preds_test = rf_model1.predict(X_test)

# accuracy score from the test data
acc_test = accuracy_score(y_test, preds_test)

print(f"Train data Accuracy score: {acc_train}")
print(f"Test data Accuracy score: {acc_test}")
print(f"Gap between Train and test accuracy: {acc_train - acc_test}")

Train data Accuracy score: 1.0
Test data Accuracy score: 0.933974358974359
Gap between Train adn test accuracy: 0.06602564102564101


In [31]:
rf_model2 = RandomForestClassifier(
    n_estimators=20, 
    max_depth=5, 
    min_samples_leaf=10, 
    max_features=0.5
)

In [32]:
def modelTrain(model, X_train, y_train, X_test, y_test):
    # fit the model with the train data
    model.fit(X_train, y_train)

    # prediction from training data
    preds_train = model.predict(X_train)

    # accuracy score from the traing data
    acc_train = accuracy_score(y_train, preds_train)

    # prediction from test data
    preds_test = model.predict(X_test)

    # accuracy score from the test data
    acc_test = accuracy_score(y_test, preds_test)

    print(f"Train data Accuracy score: {acc_train}")
    print(f"Test data Accuracy score: {acc_test}")
    print(f"Gap between Train and test accuracy: {acc_train - acc_test}")

In [33]:
modelTrain(rf_model2,X_train, y_train, X_test, y_test)

Train data Accuracy score: 0.7513227513227513
Test data Accuracy score: 0.7262820512820513
Gap between Train and test accuracy: 0.025040700040700026
